### ISSUE TITLE
Can't seem to get Working Set limiting working for Anno 1800

### ISSUE THREAD

**Issue by Nicholas-Steel** [lowleveldesign/process-governor#27]
If I try it with a BAT file and run the BAT file as an administrator the game closes almost immediately, if I try it with the Install parameter the game will persist as a Process but in a broken state with very little RAM used and no way to close it due to a lack of permissions to interact with the process.

The game seems to use around 7GB's of Active Private Working Set and will exhaust all remaining RAM as Working Set before commencing gameplay.  My goal was to limit the games Working Set so that there was around 1GB's of RAM left for other tasks on the computer to utilize/minimize the amount of paging needed when multi-tasking (similar to what the Xbox One, Xbox Series, PS4 and PS5 gaming consoles do).

I have 16GB's of RAM.

procgov64.exe -m=15G Anno1800.exe
"D:\Anno 1800\Bin\Win64\procgov64.exe" --maxws=9G --minws=9G Anno1800.exe (elevated command prompt)

Both of these will launch the game, however the game is not constrained in memory usage. Adding the -r parameter causes the game to not open.

Windows 10 21H1

**Comment by Nicholas-Steel** [comment]
Hmmm, I'm starting to think the -r parameter is breaking it. I would've thought it would be neccessary but maybe it isn't? The game launches Ubisoft Connect if it isn't already running and then Ubisoft Connect re-launches the game with the game actually opening this time.

**Comment by Nicholas-Steel** [comment]
procgov64.exe -m=15G Anno1800.exe
 "D:\Anno 1800\Bin\Win64\procgov64.exe" --maxws=9G --minws=9G Anno1800.exe (elevated command prompt)
 
 Both of these will launch the game, however the game is not constrained in memory usage.

**Comment by lowleveldesign** [comment]
The -r option propagates the limits to child processes. If you skip it, the child process will not be constrained and that explains why you see the game running. Could you please record a [procmon](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) trace with and without the -r option and share it with me? I would like to see what processes are launched. There is a chance too that the game breaks because it does not have enough free memory when run under procmon.

**Comment by Nicholas-Steel** [comment]
Does Procgov prevent a program from spilling over in to Swap/Page File? I was thinking ProcGov only controlled the amount of physical RAM the program would see as available and the influenced program would naturally start using the Paging File once it runs out of the RAM it thinks is available (assuming there is enough available RAM for the critical portions of the program to occupy).

**Comment by lowleveldesign** [comment]
You're right that when you limit the working set, procgov will apply a constraint only on the amount of used physical memory by the process. However, this still could cause problems in the target process. Btw., are there any logs in the Event Viewer about the error?

For the procmon, let's first enable only the process/thread activity: 
![image](https://user-images.githubusercontent.com/6109268/133774762-ace265e5-2dca-4d10-971e-cc3b53b38e4f.png)

**Comment by Nicholas-Steel** [comment]
Gonna take a moment to compress the data.

**Comment by Nicholas-Steel** [comment]
You can download the traces with and without -r here https://mega.nz/file/bt4ADQBL#JH_UsX2VicCszx6oG3bGR8IHWnHmlc34oRptQrHP33Y

**Comment by lowleveldesign** [comment]
Please also upload or copy-paste the G:\Ubisoft Games\logs\aux_r1_log.txt log file (after running with -r option). Make sure it does not contain anything private. I see that Anno1800.exe can't start UbisoftGameLauncher.exe and it logs something to this file. Maybe it's the error message :)

**Comment by Nicholas-Steel** [comment]
[aux_r1_log.txt](https://github.com/lowleveldesign/process-governor/files/7189590/aux_r1_log.txt)
I doubt this will be helpful but here.

**Comment by lowleveldesign** [comment]
Actually, the error helps a lot. The 1450 error means `ERROR_NO_SYSTEM_RESOURCES`. I just noticed that you set the `--minws` to a very high value. I'm unsure how setting minimum working set impacts the child processes. Could you try setting minws to a very low values, for example, `-r --minws 1M --maxws 9G`? Does it change anything?

**Comment by Nicholas-Steel** [comment]
I did try that before, but I kept getting an error until I made the min and max match.

**Comment by lowleveldesign** [comment]
Hmm, that's strange. I was able to reproduce the error 1450 with notepad when setting `--minws 8G --maxws 8G` and starting a child process. But when I switched to `--minws 1M --maxws 8G` everything worked. Could you specify what's the procgov error when you set this small --minws value?

**Comment by Nicholas-Steel** [comment]
You're right, it works. I'm thinking I may have done something dumb and set the _Maximum_ to 1MB and Minimum to 9GB when originally testing.

Thank you very much for the assistance, it's been greatly appreciated.

**Comment by Nicholas-Steel** [comment]
What does Error Code 5 mean? It's added to the "game_starter_log.txt" in Ubisoft Connect's log folder when I try to apply any limit today. Yesterday and day before that it was working fine.

**Comment by lowleveldesign** [comment]
It's access denied which usually indicates that the process does  not have access to a file. I don't think it's related to procgov.

**Comment by Nicholas-Steel** [comment]
I just make a Batch file with this content

```
"D:\Anno 1800\Bin\Win64\procgov64.exe" -r --minws=1M --maxws=9G Anno1800.exe
pause
```

and run it as an Administrator. It did previously work.

**Comment by lowleveldesign** [comment]
Try running procmon again but with all events enabled and look for ACCESS DENIED results. They should point you to the erroneous actions.

**Comment by Nicholas-Steel** [comment]
[Logfile.zip](https://github.com/lowleveldesign/process-governor/files/7196462/Logfile.zip)

**Comment by lowleveldesign** [comment]
I don't see procgov running in the trace you attached. And indeed there is no access denied error reported in the logs. I know it sounds bad, but maybe you should try reinstalling the whole game.

**Comment by Nicholas-Steel** [comment]
Does what process Governor do, leave behind any kind of data on the HDD? Is there anything I can manually delete to start fresh?

**Comment by Nicholas-Steel** [comment]
Okay well while I still believe one day it worked and the next day it stopped working and no system changes were made by me and I'm the only user of this PC... I have now figured out a way to get it to work. I now have to run Ubisoft Connect as an administrator first, and then the batch file as an Administrator.
